docs/misc/str-builder-api.rst

```diff
@@ -0,0 +1,388 @@
+    PyObject *s = PyUnicode_New(size, 127);
+    Py_UCS1 *buf = PyUnicode_1BYTE_DATA(s);
+    strcpy(buf, "hello");
+
+
+.. note::
+   CPython distinguishes between ``PyUnicode_New(size, 127)`` and
+   ``PyUnicode_New(size, 255)``: in both cases the kind is
+   ``PyUnicode_1BYTE_KIND``, but the former also sets a flag to indicate that
+   the string is ASCII-only.
+
+There are cases in which you don't know the kind in advance because you are
+working on generic data. To solve the problem in addition to the raw-buffer
+API, CPython also offers an "Opaque API" to write a char inside an unicode:
+
+.. code-block:: c
+
+    int PyUnicode_WriteChar(PyObject *unicode, Py_ssize_t index, Py_UCS4 character)
+    void PyUnicode_WRITE(int kind, void *data, Py_ssize_t index, Py_UCS4 value)
+
+Note that the character to write is always ``Py_UCS4``, so
+``_WriteChar``/``_WRITE`` have logic to do something different depending on
+the kind.
+
+.. note::
+    ``_WRITE`` is a macro, and its implementation contains a ``switch(kind)``:
+    I think it is designed with the explicit goal of allowing the compiler to
+    hoist the ``switch`` outside a loop in which we repeatedly call
+    ``_WRITE``. However, it is worth noting that I could not find any code
+    using it outside CPython itself, so it's probably something which we don't
+    need to care of for HPy.
+
+
+Raw-buffer vs Opaque API
+---------------------------
+
+There are two ways to initialize a non-initialized string object:
+
+- **Raw-buffer API**: get a C pointer to the memory and fill it directly:
+  ``PyBytes_AsString``, ``PyUnicode_1BYTE_DATA``, etc.
+
+- **Opaque API**: call special functions API to fill the content, without
+  accessing the buffer directly: e.g., ``PyUnicode_WriteChar``.
+
+From the point of view of the implementation, a completely opaque API gives
+the most flexibility in terms of how to implement a builder and/or a string.
+A good example is PyPy's ``str`` type, which uses UTF-8 as the internal
+representation. A completely opaque ``HPyStrBuilder`` could allow PyPy to fill
+directly its internal UTF-8 buffer (at least in simple cases). On the other
+hand, a raw-buffer API would force PyPy to store the UCS{1,2,4} bytes in a
+temporary buffer and convert them to UTF-8 during the ``build()`` phase.
+
+On the other hand, from the point of view of the C programmer it is easier to
+have direct access the memory. This allows to:
+
+- use ``memcpy()`` to copy data into the buffer
+
+- pass the buffer directly to other C functions which write into it (e.g.,
+  ``read()``)
+
+- use standard C patterns such as ``*p++ = ...`` or similar.
+
+
+Problems and constraints
+------------------------
+
+``bytes`` and ``str`` are objects are immutable: the biggest problem of the
+current API boils down to the fact that the API allows to construct objects
+which are not fully initialized and to mutate them during a
+not-well-specificed "initialization phase".
```

Comment:
Isn't this the goal of `_PyUnicode_Ready` in [PEP 393](https://www.python.org/dev/peps/pep-0393/)?

Reply:
no, `PyUnicode_READY` is needed if you use the legacy (now deprecated) APIs to build unicode (e.g. `PyUnicode_FromUnicode`). `PyUnicode_READY` is also deprecated and will be removed in 3.12.

If you use `PyUnicode_New + PyUnicode_1BYTE_DATA` you don't need to call `PyUnicode_READY`.